src/bvh/iter.rs

```diff
@@ -0,0 +1,377 @@
+        self.stack_size += 1;
+    }
+
+    /// Pop the stack and return the node. Not guarded against underflow.
+    fn stack_pop(&mut self) -> usize {
+        self.stack_size -= 1;
+        return self.stack[self.stack_size];
+    }
+
+    /// Attempt to move to the left child of the current node.
+    fn move_left(&mut self) {
+        match self.bvh.nodes[self.node_index] {
+            BVHNode::Node {
+                child_l_index,
+                ref child_l_aabb,
+                ..
+            } => {
+                if self.ray.intersects_aabb(child_l_aabb) {
+                    self.node_index = child_l_index;
+                    self.has_node = true;
+                } else {
+                    self.has_node = false;
+                }
+            }
+            BVHNode::Leaf { .. } => {
+                self.has_node = false;
+            }
+        }
+    }
+
+    /// Attempt to move to the right child of the current node.
+    fn move_right(&mut self) {
+        match self.bvh.nodes[self.node_index] {
+            BVHNode::Node {
+                child_r_index,
+                ref child_r_aabb,
+                ..
+            } => {
+                if self.ray.intersects_aabb(child_r_aabb) {
+                    self.node_index = child_r_index;
+                    self.has_node = true;
+                } else {
+                    self.has_node = false;
+                }
+            }
+            BVHNode::Leaf { .. } => {
+                self.has_node = false;
+            }
+        }
+    }
+}
+
+impl<'a, Shape: Bounded> Iterator for BVHIterator<'a, Shape> {
+    type Item = &'a Shape;
+
+    fn next(&mut self) -> Option<&'a Shape> {
+        loop {
+            if self.is_stack_empty() && !self.has_node {
+                // Completed traversal.
+                break;
+            }
+            if self.has_node {
+                // If we have any node, save it and attempt to move to its left child.
+                self.stack_push(self.node_index);
+                self.move_left();
+            } else {
+                // Go back up the stack and see if a node or leaf was pushed.
+                self.node_index = self.stack_pop();
+                match self.bvh.nodes[self.node_index] {
+                    BVHNode::Node { .. } => {
+                        // If a node was pushed, now attempt to move to its right child.
+                        self.move_right();
+                    }
+                    BVHNode::Leaf { shape_index, .. } => {
+                        // We previously pushed a leaf node. This is the "visit" of the in-order traverse.
+                        // Next time we call next we try to pop the stack again.
+                        self.has_node = false;
+                        return Some(&self.shapes[shape_index]);
+                    }
+                }
+            }
+        }
+        return None;
+    }
+}
+
+// Copy of part of the BH testing in testbase.
```

Comment:
Any chance you could unify this back to testbase? I think we're going to see less duplication that way. See, for instance, [here](https://github.com/svenstaro/bvh/blob/master/src/bvh/optimization.rs#L520) where we re-use a lot of stuff from testbase. This might be helpful.

Reply:
Yes, I tried actually. The stumbling block for me was testbase functions all talk to BoundingHierarchy, but my traverse_iterator() is only on BVH. If you have a suggestion of how to express traverse_iterator on BoundingHierarchy I would absolutely implement it there (and for flat_bvh etc.) so we don't have this weird BVH-only traversal. But when I tried to add it on BoundingHierarchy, the rust compiler wanted me to suddenly throw "dyn"s everywhere. My lack of Rust knowledge let me down.

Reply:
`dyn` is usually only associated with [trait objects](https://doc.rust-lang.org/book/ch17-02-trait-objects.html) so Rust somehow thought you were going for a trait object rather than an implementation. Maybe your syntax was slightly off?

Reply:
The obstacle I hit was: `traverse_iterator` on `BoundingHierarchy` must return _something_... It can't return `BVHIterator`. Like a base `TraverseIterator`... But, as soon as I return a trait, rust warns me:
warning: trait objects without an explicit `dyn` are deprecated

I guess that's just how it is though, without dynamic dispatch the runtime doesn't know what kind of iterator we're dealing with when looping over the results of `traverse_iterator` on a `BoundingHierarchy` 🤷‍♂️  .. Now I just need to figure out how to expose a generic iterator in `BoundingHierarchy` and the actual specialized iterators for `BVH`/`FlatBVH`

I like `TraverseIterator` and `BVHTraverseIterator` by the way. Adding "Traverse" feels more descriptive. Should I follow this naming?

Reply:
Sure, go for it.

Reply:
Well, I went down the rabbit hole with `BoundingHierarchy` having a `traverse_iterator() -> Iterator<Shape: BHShape>` and I landed with a solution which depends on `#![feature(generic_associated_types)]` .. Hah. So the only way I could expose an iterator on `BoundingHierarchy` without lots of boxing and `dyn` is through a type in the `BoundingHierarchy`. Specifically this construction:

```
    type Iter<Shape: BHShape>: Iterator<Item = Shape>;
    fn traverse_iterator<'a, Shape: BHShape>(&'a self, ray: &Ray, shapes: &'a [Shape]) -> Self::Iter<Shape>;
```

This allows implementations like `BVH` to do:
```
    type Iter<Shape: BHShape> = BVHTraverseIterator<Item = Shape>;
    fn traverse_iterator<'a, Shape: BHShape>(&'a self, ray: &Ray, shapes: &'a [Shape]) -> Self::Iter<Shape>;
```

Unfortunately, it depends on `GAT` which you may not want to depend on yet...
```
error[E0658]: generic associated types are unstable
   --> src/bounding_hierarchy.rs:168:5
    |
168 |     type Iter<Shape: BHShape>: Iterator<Item = Shape>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #44265 <https://github.com/rust-lang/rust/issues/44265> for more information
    = help: add `#![feature(generic_associated_types)]` to the crate attributes to enable
```

So I think I'd just keep traverse_iterator() on BVH only for now. Anyway I'll try to come up with some better way of sharing code with testbase :)

Reply:
That was quite the adventure. I agree with your suggestion to keep the `traverse_iterator()` separate for the time being. We can always improve on this.

Reply:
Sounds good. I've just removed as much of the duplicated testbase code as I could. It's still a little verbose with its own `traverse_some_bvh` etc. But still much cleaner.